### PR TITLE
Add Tor Outbound Support 

### PR DESCRIPTION
**Pull request by MeshCollider**
Replaces #389, picking up where Crypt-iQ left off. Most review comments have been addressed, one thing from the original PR description that has not been addressed is:

> Support for Hidden Services (v2/v3), or support for a bridge mode is not yet implemented

~The other comment which hasn't yet been addressed is https://github.com/lightningnetwork/lnd/pull/389#discussion_r147606429 because I'm still unsure on a clean way to do it~

### CODE REVIEW

File: channeldb/channel.go [hunk 1]
```diff
@@ -389,6 +395,12 @@ type OpenChannel struct {
 	sync.RWMutex
 }
 
+// SetResolver sets resolveTCP to a resolver other than the default
```
Comment (Roasbeef) [inline channeldb/channel.go]: This isn't currently used in the PR. I now remember it's original purpose: to ensure that any DNS resolutions we performed underlying deserialization didn't also leak to the clearnet if we were in tor mode. I think a better solution is to _remove any sort of DNS queries_ during deserialization all together. This would mean modifying the on-disk storage format of the addresses to _not_ just store the string, but store the raw representation. 

So in short, we abstract out [this logic](https://github.com/lightningnetwork/lnd/blob/master/channeldb/graph.go#L1740), so it can be used for both `LinkNodes` as well as `LightningNodes`.

File: config.go [hunk 2]
```diff
@@ -284,6 +294,51 @@ func loadConfig() (*config, error) {
...
+			str := "%s: The tor socks5 port must be between 1024 and 65535"
+			err := fmt.Errorf(str, funcName)
+			fmt.Fprintln(os.Stderr, err)
+			fmt.Fprintln(os.Stderr, usageMessage)
+			return nil, err
+		}
+
+		// If ExternalIPs is set, throw an error since we cannot
+		// listen for incoming connections via Tor's SOCKS5 proxy.
+		if len(cfg.ExternalIPs) != 0 {
```
Comment (Roasbeef) [inline config.go]: Ah nvm I see that this is actually there from the existing set of commits!

**Comment by MeshCollider** [comment]
Removed [WIP] tag after making the requested changes to channeldb address serialization to remove the address resolution from that package

**Comment by Crypt-iQ** [comment]
Took a look at the removal of channeldb address resolution - I was a little stumped on how to do that cleanly as well, but it seems like you did it! Good job!

**Comment by Roasbeef** [comment]
Just hit this locally while testing: 
```
2018-02-02 16:00:10.321 [DBG] PEER: Received MsgOpenChannel(temp_chan_id=97a03fb44b7c5ee37a15b09ad1cdbc68c4a8c7db9da3651445d43959a523975a, chain=000000000933ea01ad0ee984209779baaec3ced90fa3f408719526f8d77f4943, csv=1922, amt=0.16 BTC, push_amt=8000000000 mSAT, reserve=0.0016 BTC, flags=1) from 159.203.125.125:9735
2018-02-02 16:00:10.323 [INF] FNDG: Recv'd fundingRequest(amt=0.16 BTC, push=8000000000 mSAT, delay=1922, pendingId=97a03fb44b7c5ee37a15b09ad1cdbc68c4a8c7db9da3651445d43959a523975a) from peer(02fa77e0f4ca666f7d158c4bb6675d1436e339903a9feeeaacbd6e55021b98e7ee)
panic: interface conversion: net.Addr is *socks.ProxiedAddr, not *net.TCPAddr

goroutine 38 [running]:
main.(*fundingManager).handleFundingOpen(0xc4322c8000, 0xc432386750)
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/fundingmanager.go:839 +0x1ecf
main.(*fundingManager).reservationCoordinator(0xc4322c8000)
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/fundingmanager.go:703 +0x278
created by main.(*fundingManager).Start
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/fundingmanager.go:551 +0xe06
```

**Comment by Roasbeef** [comment]
It's due to the fact that rn, we have an unchecked type assertion within the `fundingManager`. We use this to note the address of the node that we're initiating the channel with (or the one that initiated it with us). It seems that when accepting an inbound funding request (established a session with the faucet over Tor), this address is actually a `*socks.ProxiedAddr` not a `*net.TCPAddr`. 

The issue is that this eventually leads to a call of `SyncPending`, which accepts a `*net.TCPAddr` atm. I think we can simply just modify this to actually be a `net.Addr` instead, and update the serialization accordingly.

**Comment by Roasbeef** [comment]
If this diff is applied, everything works properly:
```diff
diff --git a/channeldb/addr.go b/channeldb/addr.go
index daa61a8b..0599b3af 100644
--- a/channeldb/addr.go
+++ b/channeldb/addr.go
@@ -3,6 +3,8 @@ package channeldb
 import (
 	"io"
 	"net"
+
+	"github.com/btcsuite/go-socks/socks"
 )
 
 // addressType specifies the network protocol and version that should be used
@@ -67,36 +69,69 @@ func deserializeAddr(r io.Reader) (net.Addr, error) {
 	return address, nil
 }
 
+func encodeTcpAddr(w io.Writer, addr *net.TCPAddr) error {
+	var scratch [16]byte
+
+	if addr.IP.To4() != nil {
+		scratch[0] = uint8(tcp4Addr)
+		if _, err := w.Write(scratch[:1]); err != nil {
+			return err
+		}
+
+		copy(scratch[:4], addr.IP.To4())
+		if _, err := w.Write(scratch[:4]); err != nil {
+			return err
+		}
+
+	} else {
+		scratch[0] = uint8(tcp6Addr)
+		if _, err := w.Write(scratch[:1]); err != nil {
+			return err
+		}
+
+		copy(scratch[:], addr.IP.To16())
+		if _, err := w.Write(scratch[:]); err != nil {
+			return err
+		}
+	}
+
+	byteOrder.PutUint16(scratch[:2], uint16(addr.Port))
+	if _, err := w.Write(scratch[:2]); err != nil {
+		return err
+	}
+
+	return nil
+}
+
 // serializeAddr serializes an address into a raw byte representation so it
 // can be deserialized without requiring address resolution
 func serializeAddr(w io.Writer, address net.Addr) error {
-	var scratch [16]byte
 
-	if address.Network() == "tcp" {
-		if address.(*net.TCPAddr).IP.To4() != nil {
-			scratch[0] = uint8(tcp4Addr)
-			if _, err := w.Write(scratch[:1]); err != nil {
-				return err
-			}
-			copy(scratch[:4], address.(*net.TCPAddr).IP.To4())
-			if _, err := w.Write(scratch[:4]); err != nil {
-				return err
-			}
-		} else {
-			scratch[0] = uint8(tcp6Addr)
-			if _, err := w.Write(scratch[:1]); err != nil {
-				return err
-			}
-			copy(scratch[:], address.(*net.TCPAddr).IP.To16())
-			if _, err := w.Write(scratch[:]); err != nil {
-				return err
-			}
+	switch addr := address.(type) {
+	case *net.TCPAddr:
+		return encodeTcpAddr(w, addr)
+
+	// If this is a proxied address (due to the connection being
+	// established over a SOCKs proxy, then we'll convert it into its
+	// corresponding TCP address.
+	case *socks.ProxiedAddr:
+		// If we can't parse the host as an IP (though we should be
+		// able to at this point), then we'll skip this address all
+		// together.
+		//
+		// TODO(roasbeef): would be nice to be able to store hosts
+		// tho...
+		ip := net.ParseIP(addr.Host)
+		if ip == nil {
+			return nil
 		}
-		byteOrder.PutUint16(scratch[:2],
-			uint16(address.(*net.TCPAddr).Port))
-		if _, err := w.Write(scratch[:2]); err != nil {
-			return err
+
+		tcpAddr := &net.TCPAddr{
+			IP:   ip,
+			Port: addr.Port,
 		}
+		return encodeTcpAddr(w, tcpAddr)
+
 	}
 
 	return nil
diff --git a/channeldb/channel.go b/channeldb/channel.go
index 8e246e3e..d9d56c96 100644
--- a/channeldb/channel.go
+++ b/channeldb/channel.go
@@ -623,7 +623,7 @@ func fetchOpenChannel(chanBucket *bolt.Bucket,
 //
 // TODO(roasbeef): addr param should eventually be a lnwire.NetAddress type
 // that includes service bits.
-func (c *OpenChannel) SyncPending(addr *net.TCPAddr, pendingHeight uint32) error {
+func (c *OpenChannel) SyncPending(addr net.Addr, pendingHeight uint32) error {
 	c.Lock()
 	defer c.Unlock()
 
diff --git a/channeldb/nodes.go b/channeldb/nodes.go
index 9274bc08..57626e76 100644
--- a/channeldb/nodes.go
+++ b/channeldb/nodes.go
@@ -56,7 +56,7 @@ type LinkNode struct {
 	// authenticated connection for the stored identity public key.
 	//
 	// TODO(roasbeef): also need to support hidden service addrs
-	Addresses []*net.TCPAddr
+	Addresses []net.Addr
 
 	db *DB
 }
@@ -64,13 +64,13 @@ type LinkNode struct {
 // NewLinkNode creates a new LinkNode from the provided parameters, which is
 // backed by an instance of channeldb.
 func (db *DB) NewLinkNode(bitNet wire.BitcoinNet, pub *btcec.PublicKey,
-	addr *net.TCPAddr) *LinkNode {
+	addr net.Addr) *LinkNode {
 
 	return &LinkNode{
 		Network:     bitNet,
 		IdentityPub: pub,
 		LastSeen:    time.Now(),
-		Addresses:   []*net.TCPAddr{addr},
+		Addresses:   []net.Addr{addr},
 		db:          db,
 	}
 }
@@ -267,13 +267,13 @@ func deserializeLinkNode(r io.Reader) (*LinkNode, error) {
 	}
 	numAddrs := byteOrder.Uint32(buf[:4])
 
-	node.Addresses = make([]*net.TCPAddr, numAddrs)
+	node.Addresses = make([]net.Addr, numAddrs)
 	for i := uint32(0); i < numAddrs; i++ {
 		addr, err := deserializeAddr(r)
 		if err != nil {
 			return nil, err
 		}
-		node.Addresses[i] = addr.(*net.TCPAddr)
+		node.Addresses[i] = addr
 	}
 
 	return node, nil
diff --git a/fundingmanager.go b/fundingmanager.go
index 8fd97eb7..af01e88b 100644
--- a/fundingmanager.go
+++ b/fundingmanager.go
@@ -12,6 +12,8 @@ import (
 
 	"golang.org/x/crypto/salsa20"
 
+	"net"
+
 	"github.com/boltdb/bolt"
 	"github.com/davecgh/go-spew/spew"
 	"github.com/go-errors/errors"
@@ -26,7 +28,6 @@ import (
 	"github.com/roasbeef/btcd/chaincfg/chainhash"
 	"github.com/roasbeef/btcd/wire"
 	"github.com/roasbeef/btcutil"
-	"net"
 )
 
 const (
@@ -836,7 +837,7 @@ func (f *fundingManager) handleFundingOpen(fmsg *fundingOpenMsg) {
 	chainHash := chainhash.Hash(msg.ChainHash)
 	reservation, err := f.cfg.Wallet.InitChannelReservation(amt, 0,
 		msg.PushAmount, btcutil.Amount(msg.FeePerKiloWeight), 0,
-		fmsg.peerAddress.IdentityKey, fmsg.peerAddress.Address.(*net.TCPAddr),
+		fmsg.peerAddress.IdentityKey, fmsg.peerAddress.Address,
 		&chainHash, msg.ChannelFlags)
 	if err != nil {
 		fndgLog.Errorf("Unable to initialize reservation: %v", err)
@@ -1209,8 +1210,8 @@ func (f *fundingManager) handleFundingCreated(fmsg *fundingCreatedMsg) {
 	// send it to the ChainArbitrator so it can watch for any on-chain
 	// actions during this final confirmation stage.
 	if err := f.cfg.WatchNewChannel(completeChan); err != nil {
-		fndgLog.Error("Unable to send new ChannelPoint(%v) for "+
-			"arbitration", fundingOut)
+		fndgLog.Errorf("Unable to send new ChannelPoint(%v) for "+
+			"arbitration: %v", fundingOut, err)
 	}
 
 	// Create an entry in the local discovery map so we can ensure that we
@@ -1348,8 +1349,8 @@ func (f *fundingManager) handleFundingSigned(fmsg *fundingSignedMsg) {
 	// watch for any on-chin actions before the channel has fully
 	// confirmed.
 	if err := f.cfg.WatchNewChannel(completeChan); err != nil {
-		fndgLog.Error("Unable to send new ChannelPoint(%v) for "+
-			"arbitration", fundingPoint)
+		fndgLog.Errorf("Unable to send new ChannelPoint(%v) for "+
+			"arbitration: %v", fundingPoint, err)
 	}
 
 	fndgLog.Infof("Finalizing pendingID(%x) over ChannelPoint(%v), "+
diff --git a/lnwallet/reservation.go b/lnwallet/reservation.go
index 51f91298..4793ac31 100644
--- a/lnwallet/reservation.go
+++ b/lnwallet/reservation.go
@@ -110,7 +110,7 @@ type ChannelReservation struct {
 	theirContribution *ChannelContribution
 
 	partialState *channeldb.OpenChannel
-	nodeAddr     *net.TCPAddr
+	nodeAddr     net.Addr
 
 	// The ID of this reservation, used to uniquely track the reservation
 	// throughout its lifetime.
diff --git a/lnwallet/wallet.go b/lnwallet/wallet.go
index e96e20a6..10090f01 100644
--- a/lnwallet/wallet.go
+++ b/lnwallet/wallet.go
@@ -82,10 +82,10 @@ type initFundingReserveMsg struct {
 	// with.
 	nodeID *btcec.PublicKey
 
-	// nodeAddr is the IP address plus port that we used to either
-	// establish or accept the connection which led to the negotiation of
-	// this funding workflow.
-	nodeAddr *net.TCPAddr
+	// nodeAddr is the address port that we used to either establish or
+	// accept the connection which led to the negotiation of this funding
+	// workflow.
+	nodeAddr net.Addr
 
 	// fundingAmount is the amount of funds requested for this channel.
 	fundingAmount btcutil.Amount
@@ -452,7 +452,7 @@ out:
 func (l *LightningWallet) InitChannelReservation(
 	capacity, ourFundAmt btcutil.Amount, pushMSat lnwire.MilliSatoshi,
 	commitFeePerKw, fundingFeePerWeight btcutil.Amount,
-	theirID *btcec.PublicKey, theirAddr *net.TCPAddr,
+	theirID *btcec.PublicKey, theirAddr net.Addr,
 	chainHash *chainhash.Hash, flags lnwire.FundingFlag) (*ChannelReservation, error) {
 
 	errChan := make(chan error, 1)
diff --git a/server.go b/server.go
index 08de6e59..66c0b68c 100644
--- a/server.go
+++ b/server.go
@@ -815,7 +815,7 @@ func (s *server) genNodeAnnouncement(
 
 type nodeAddresses struct {
 	pubKey    *btcec.PublicKey
-	addresses []*net.TCPAddr
+	addresses []net.Addr
 }
 
 // establishPersistentConnections attempts to establish persistent connections
@@ -838,9 +838,13 @@ func (s *server) establishPersistentConnections() error {
 	}
 	for _, node := range linkNodes {
 		for _, address := range node.Addresses {
-			if address.Port == 0 {
-				address.Port = defaultPeerPort
+			switch addr := address.(type) {
+			case *net.TCPAddr:
+				if addr.Port == 0 {
+					addr.Port = defaultPeerPort
+				}
 			}
+
 		}
 		pubStr := string(node.IdentityPub.SerializeCompressed())
 
@@ -872,18 +876,24 @@ func (s *server) establishPersistentConnections() error {
 		// list of addresses we'll connect to. If there are duplicates
 		// that have different ports specified, the port from the
 		// channel graph should supersede the port from the link node.
-		var addrs []*net.TCPAddr
+		var addrs []net.Addr
 		linkNodeAddrs, ok := nodeAddrsMap[pubStr]
 		if ok {
 			for _, lnAddress := range linkNodeAddrs.addresses {
+				lnAddrTcp, ok := lnAddress.(*net.TCPAddr)
+				if !ok {
+					continue
+				}
+
 				var addrMatched bool
 				for _, polAddress := range policy.Node.Addresses {
 					polTCPAddr, ok := polAddress.(*net.TCPAddr)
-					if ok && polTCPAddr.IP.Equal(lnAddress.IP) {
+					if ok && polTCPAddr.IP.Equal(lnAddrTcp.IP) {
 						addrMatched = true
 						addrs = append(addrs, polTCPAddr)
 					}
 				}
+
 				if !addrMatched {
 					addrs = append(addrs, lnAddress)
 				}

```

**Comment by Roasbeef** [comment]
We'll detect a proxied address, and convert it into a regular TCP address. The downside of this approach is that we miss host information.

**Comment by MeshCollider** [comment]
Patch applied, thank you @Roasbeef!

**Comment by Crypt-iQ** [comment]
Once this is merged in, I want to work on hidden services.  @MeshCollider is that ok?

**Comment by MeshCollider** [comment]
@Crypt-iQ sure, go for it :)

**Comment by Roasbeef** [comment]
Working on a follow up PR to this which adds documentation, along with support for stream isolation.
